File: generate_faq.py
import sqlite3
import json

DB_FILE = "d:/roadwatch_desktop/data/contracts.db"
conn = sqlite3.connect(DB_FILE)
cursor = conn.cursor()

def format_inr(value):
    if not value: return "N/A"
    if value >= 10000000: return f"₹{value / 10000000:.2f} Cr"
    return f"₹{value / 100000:.2f} L"

faqs = []

# Q1: Overall Stats
cursor.execute("SELECT COUNT(*), SUM(contract_value), AVG(contract_value) FROM contracts")
count, total, avg = cursor.fetchone()
faqs.append({
    "question": "What is the overall highway infrastructure investment overview?",
    "answer": "Here is the high-level overview of all highway contracts tracked in the database.",
    "stats": {
        "Total Contracts": str(count),
        "Total Investment": format_inr(total),
        "Average Contract": format_inr(avg)
    }
})

# Q2: Top 10 States
cursor.execute("SELECT state, COUNT(*), SUM(contract_value) as val FROM contracts WHERE state != 'Other' AND state != '' GROUP BY state ORDER BY val DESC LIMIT 10")
rows = cursor.fetchall()
table_rows = [[r[0], str(r[1]), format_inr(r[2])] for r in rows]
faqs.append({
    "question": "Show me the top 10 states by highway investment.",
    "answer": "Here is the breakdown of the top 10 states receiving the highest highway infrastructure investments.",
    "table": {
        "columns": ["State", "Number of Contracts", "Total Value Awarded"],
        "rows": table_rows
    }
})

# Q3: Top 10 Contractors
cursor.execute("SELECT selected_bidder, COUNT(*), SUM(contract_value) as val FROM contracts WHERE selected_bidder != '' GROUP BY selected_bidder ORDER BY val DESC LIMIT 10")
rows = cursor.fetchall()
table_rows = []
for r in rows:
    bidder = r[0]
    if bidder and ',' in bidder:
        bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([bidder, str(r[1]), format_inr(r[2])])

faqs.append({
    "question": "Who are the top 10 highway contractors nationwide?",
    "answer": "These are the top 10 contractors based on total awarded contract value.",
    "table": {
        "columns": ["Contractor", "Contracts Won", "Total Value Awarded"],
        "rows": table_rows
    }
})

# Q4: Largest Contracts
cursor.execute("SELECT tender_ref_no, selected_bidder, state, contract_value FROM contracts ORDER BY contract_value DESC LIMIT 10")
rows = cursor.fetchall()
table_rows = []
for r in rows:
    bidder = r[1]
    if bidder and ',' in bidder:
        bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([r[0][:20]+"...", bidder, r[2] or "N/A", format_inr(r[3])])

faqs.append({
    "question": "What are the top 10 largest individual highway contracts?",
    "answer": "Here are the 10 most massive single highway infrastructure projects awarded.",
    "table": {
        "columns": ["Tender Ref", "Contractor", "State", "Contract Value"],
        "rows": table_rows
    }
})

# Q5: Most Competitive Tenders
cursor.execute("SELECT tender_ref_no, state, bids_received, contract_value FROM contracts WHERE bids_received > 0 ORDER BY bids_received DESC LIMIT 10")
rows = cursor.fetchall()
table_rows = [[r[0][:20]+"...", r[1] or "N/A", str(r[2]), format_inr(r[3])] for r in rows]
faqs.append({
    "question": "Which tenders received the highest number of bids?",
    "answer": "These are the most highly competitive highway tenders, ranked by the number of bids received.",
    "table": {
        "columns": ["Tender Ref", "State", "Bids Received", "Contract Value"],
        "rows": table_rows
    }
})

# Q6: Uttar Pradesh Stats
cursor.execute("SELECT COUNT(*), SUM(contract_value) FROM contracts WHERE state LIKE '%Uttar Pradesh%'")
count, total = cursor.fetchone()
cursor.execute("SELECT selected_bidder, SUM(contract_value) as val FROM contracts WHERE state LIKE '%Uttar Pradesh%' AND selected_bidder != '' GROUP BY selected_bidder ORDER BY val DESC LIMIT 5")
top_bidders = cursor.fetchall()
table_rows = []
for r in top_bidders:
    bidder = r[0]
    if bidder and ',' in bidder: bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([bidder, format_inr(r[1])])

faqs.append({
    "question": "Detailed spending and top contractors in Uttar Pradesh.",
    "answer": f"Uttar Pradesh has {count} contracts totaling {format_inr(total)}. Here are the top contractors in the state:",
    "table": {
        "columns": ["Contractor", "Total Awarded in UP"],
        "rows": table_rows
    }
})

# Q7: Maharashtra Stats
cursor.execute("SELECT COUNT(*), SUM(contract_value) FROM contracts WHERE state LIKE '%Maharashtra%'")
count, total = cursor.fetchone()
cursor.execute("SELECT selected_bidder, SUM(contract_value) as val FROM contracts WHERE state LIKE '%Maharashtra%' AND selected_bidder != '' GROUP BY selected_bidder ORDER BY val DESC LIMIT 5")
top_bidders = cursor.fetchall()
table_rows = []
for r in top_bidders:
    bidder = r[0]
    if bidder and ',' in bidder: bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([bidder, format_inr(r[1])])

faqs.append({
    "question": "Detailed spending and top contractors in Maharashtra.",
    "answer": f"Maharashtra has {count} contracts totaling {format_inr(total)}. Here are the top contractors in the state:",
    "table": {
        "columns": ["Contractor", "Total Awarded in MH"],
        "rows": table_rows
    }
})

# Q8: Contractors by Volume (Number of Contracts)
cursor.execute("SELECT selected_bidder, COUNT(*) as c, SUM(contract_value) FROM contracts WHERE selected_bidder != '' GROUP BY selected_bidder ORDER BY c DESC LIMIT 10")
rows = cursor.fetchall()
table_rows = []
for r in rows:
    bidder = r[0]
    if bidder and ',' in bidder: bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([bidder, str(r[1]), format_inr(r[2])])

faqs.append({
    "question": "Which contractors have won the most individual contracts (volume)?",
    "answer": "These contractors have secured the highest volume of individual highway contracts.",
    "table": {
        "columns": ["Contractor", "Number of Contracts", "Total Value"],
        "rows": table_rows
    }
})

# Q9: Lowest Value Contracts
cursor.execute("SELECT tender_ref_no, state, selected_bidder, contract_value FROM contracts WHERE contract_value > 0 ORDER BY contract_value ASC LIMIT 10")
rows = cursor.fetchall()
table_rows = []
for r in rows:
    bidder = r[2]
    if bidder and ',' in bidder: bidder = bidder.split(',')[0].strip() + " (and others)"
    table_rows.append([r[0][:20]+"...", r[1] or "N/A", bidder or "N/A", format_inr(r[3])])

faqs.append({
    "question": "Show me the smallest/lowest value highway contracts.",
    "answer": "Here are the smallest recorded infrastructure contracts in the registry.",
    "table": {
        "columns": ["Tender Ref", "State", "Contractor", "Contract Value"],
        "rows": table_rows
    }
})

# Q10: Contracts by Year
cursor.execute("SELECT year, COUNT(*), SUM(contract_value) as val FROM contracts WHERE year IS NOT NULL AND year > 2000 GROUP BY year ORDER BY year DESC")
rows = cursor.fetchall()
table_rows = [[str(r[0]), str(r[1]), format_inr(r[2])] for r in rows]
faqs.append({
    "question": "What is the breakdown of highway investments by year?",
    "answer": "This table shows the progression of highway infrastructure spending categorized by year.",
    "table": {
        "columns": ["Year", "Number of Contracts", "Total Value"],
        "rows": table_rows
    }
})

with open("d:/roadwatch_desktop/core/faq_data.py", "w", encoding="utf-8") as f:
    f.write("FAQ_DATA = ")
    f.write(json.dumps(faqs, indent=4))

File: scripts/generateFaq.js
/**
 * generateFaq.js
 *
 * Mirrors the 10 FAQ queries from generate_faq.py but reads from
 * contracts_store.json (JSON) instead of SQLite. Writes faq_data.json
 * into the project root so the chat API can load it at runtime.
 */

const fs = require('fs');
const path = require('path');

const CONTRACTS_FILE = path.join(__dirname, '..', 'contracts_store.json');
const OUTPUT_FILE = path.join(__dirname, '..', 'faq_data.json');

function formatInr(value) {
  if (!value) return 'N/A';
  if (value >= 10000000) return `₹${(value / 10000000).toFixed(2)} Cr`;
  if (value >= 100000) return `₹${(value / 100000).toFixed(2)} L`;
  return `₹${value.toLocaleString('en-IN')}`;
}

function truncateBidder(name) {
  if (!name) return 'N/A';
  if (name.includes(',')) return name.split(',')[0].trim() + ' (and others)';
  return name;
}

// Load contracts
let contracts = [];
try {
  const raw = fs.readFileSync(CONTRACTS_FILE, 'utf8');
  contracts = JSON.parse(raw);
  console.log(`Loaded ${contracts.length} contracts from contracts_store.json`);
} catch (e) {
  console.error('Failed to read contracts_store.json:', e.message);
  process.exit(1);
}

const faqs = [];

// ── Q1: Overall Stats ──────────────────────────────────────────────────
const totalValue = contracts.reduce((s, c) => s + (c.contractValue || 0), 0);
const avgValue = contracts.length ? totalValue / contracts.length : 0;
faqs.push({
  id: 'overall_stats',
  question: 'What is the overall highway infrastructure investment overview?',
  keywords: ['overall', 'overview', 'investment', 'infrastructure', 'total', 'how much', 'highway spending'],
  answer: 'Here is the high-level overview of all highway contracts tracked in the database.',
  stats: {
    'Total Contracts': String(contracts.length),
    'Total Investment': formatInr(totalValue),
    'Average Contract': formatInr(avgValue),
  },
});

// ── Q2: Top 10 States ──────────────────────────────────────────────────
const stateAgg = {};
contracts.forEach((c) => {
  const st = c.state || '';
  if (!st || st === 'Other') return;
  if (!stateAgg[st]) stateAgg[st] = { count: 0, value: 0 };
  stateAgg[st].count++;
  stateAgg[st].value += c.contractValue || 0;
});
const topStates = Object.entries(stateAgg)
  .sort((a, b) => b[1].value - a[1].value)
  .slice(0, 10);
faqs.push({
  id: 'top_states',
  question: 'Show me the top 10 states by highway investment.',
  keywords: ['top states', 'states by investment', 'state ranking', 'which states', 'state spending', 'state wise'],
  answer: 'Here is the breakdown of the top 10 states receiving the highest highway infrastructure investments.',
  table: {
    columns: ['State', 'Number of Contracts', 'Total Value Awarded'],
    rows: topStates.map(([st, d]) => [st, String(d.count), formatInr(d.value)]),
  },
});

// ── Q3: Top 10 Contractors ─────────────────────────────────────────────
const bidderAgg = {};
contracts.forEach((c) => {
  const b = c.selectedBidder || '';
  if (!b) return;
  if (!bidderAgg[b]) bidderAgg[b] = { count: 0, value: 0 };
  bidderAgg[b].count++;
  bidderAgg[b].value += c.contractValue || 0;
});
const topContractors = Object.entries(bidderAgg)
  .sort((a, b) => b[1].value - a[1].value)
  .slice(0, 10);
faqs.push({
  id: 'top_contractors',
  question: 'Who are the top 10 highway contractors nationwide?',
  keywords: ['top contractors', 'biggest contractors', 'contractor ranking', 'who builds', 'largest contractors'],
  answer: 'These are the top 10 contractors based on total awarded contract value.',
  table: {
    columns: ['Contractor', 'Contracts Won', 'Total Value Awarded'],
    rows: topContractors.map(([b, d]) => [truncateBidder(b), String(d.count), formatInr(d.value)]),
  },
});

// ── Q4: Largest Contracts ──────────────────────────────────────────────
const sortedByValue = [...contracts].sort((a, b) => (b.contractValue || 0) - (a.contractValue || 0));
faqs.push({
  id: 'largest_contracts',
  question: 'What are the top 10 largest individual highway contracts?',
  keywords: ['largest contracts', 'biggest contract', 'most expensive', 'highest value', 'mega projects'],
  answer: 'Here are the 10 most massive single highway infrastructure projects awarded.',
  table: {
    columns: ['Tender Ref', 'Contractor', 'State', 'Contract Value'],
    rows: sortedByValue.slice(0, 10).map((c) => [
      (c.tenderRefNo || '').slice(0, 25) + '...',
      truncateBidder(c.selectedBidder),
      c.state || 'N/A',
      formatInr(c.contractValue),
    ]),
  },
});

// ── Q5: Most Competitive Tenders ───────────────────────────────────────
const sortedByBids = [...contracts]
  .filter((c) => c.bidsReceived > 0)
  .sort((a, b) => b.bidsReceived - a.bidsReceived);
faqs.push({
  id: 'competitive_tenders',
  question: 'Which tenders received the highest number of bids?',
  keywords: ['most bids', 'competitive', 'bidding war', 'highest bids', 'most competitive'],
  answer: 'These are the most highly competitive highway tenders, ranked by the number of bids received.',
  table: {
    columns: ['Tender Ref', 'State', 'Bids Received', 'Contract Value'],
    rows: sortedByBids.slice(0, 10).map((c) => [
      (c.tenderRefNo || '').slice(0, 25) + '...',
      c.state || 'N/A',
      String(c.bidsReceived),
      formatInr(c.contractValue),
    ]),
  },
});

// ── Q6: Uttar Pradesh Stats ────────────────────────────────────────────
function buildStateFaq(id, stateName, stateMatch, label) {
  const stateContracts = contracts.filter((c) =>
    (c.state || '').toLowerCase().includes(stateMatch)
  );
  const count = stateContracts.length;
  const total = stateContracts.reduce((s, c) => s + (c.contractValue || 0), 0);

  // top 5 contractors in this state
  const bidders = {};
  stateContracts.forEach((c) => {
    const b = c.selectedBidder || '';
    if (!b) return;
    if (!bidders[b]) bidders[b] = 0;
    bidders[b] += c.contractValue || 0;
  });
  const topBidders = Object.entries(bidders)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5);

  return {
    id,
    question: `Detailed spending and top contractors in ${stateName}.`,
    keywords: [stateName.toLowerCase(), `spending in ${stateName.toLowerCase()}`, `contractors in ${stateName.toLowerCase()}`],
    answer: `${stateName} has ${count} contracts totaling ${formatInr(total)}. Here are the top contractors in the state:`,
    table: {
      columns: ['Contractor', `Total Awarded in ${label}`],
      rows: topBidders.map(([b, v]) => [truncateBidder(b), formatInr(v)]),
    },
  };
}
faqs.push(buildStateFaq('up_stats', 'Uttar Pradesh', 'uttar pradesh', 'UP'));
faqs.push(buildStateFaq('mh_stats', 'Maharashtra', 'maharashtra', 'MH'));

// ── Q8: Contractors by Volume ──────────────────────────────────────────
const topByVolume = Object.entries(bidderAgg)
  .sort((a, b) => b[1].count - a[1].count)
  .slice(0, 10);
faqs.push({
  id: 'contractors_by_volume',
  question: 'Which contractors have won the most individual contracts (volume)?',
  keywords: ['most contracts won', 'contractor volume', 'number of contracts', 'most prolific'],
  answer: 'These contractors have secured the highest volume of individual highway contracts.',
  table: {
    columns: ['Contractor', 'Number of Contracts', 'Total Value'],
    rows: topByVolume.map(([b, d]) => [truncateBidder(b), String(d.count), formatInr(d.value)]),
  },
});

// ── Q9: Lowest Value Contracts ─────────────────────────────────────────
const sortedAsc = [...contracts]
  .filter((c) => c.contractValue > 0)
  .sort((a, b) => a.contractValue - b.contractValue);
faqs.push({
  id: 'lowest_contracts',
  question: 'Show me the smallest/lowest value highway contracts.',
  keywords: ['smallest contracts', 'lowest value', 'cheapest', 'minimum value'],
  answer: 'Here are the smallest recorded infrastructure contracts in the registry.',
  table: {
    columns: ['Tender Ref', 'State', 'Contractor', 'Contract Value'],
    rows: sortedAsc.slice(0, 10).map((c) => [
      (c.tenderRefNo || '').slice(0, 25) + '...',
      c.state || 'N/A',
      truncateBidder(c.selectedBidder),
      formatInr(c.contractValue),
    ]),
  },
});

// ── Q10: Contracts by Year ─────────────────────────────────────────────
const yearAgg = {};
contracts.forEach((c) => {
  const y = c.year;
  if (!y || y <= 2000) return;
  if (!yearAgg[y]) yearAgg[y] = { count: 0, value: 0 };
  yearAgg[y].count++;
  yearAgg[y].value += c.contractValue || 0;
});
const yearRows = Object.entries(yearAgg)
  .sort((a, b) => Number(b[0]) - Number(a[0]));
faqs.push({
  id: 'by_year',
  question: 'What is the breakdown of highway investments by year?',
  keywords: ['by year', 'yearly', 'year wise', 'annual', 'year breakdown', 'spending per year'],
  answer: 'This table shows the progression of highway infrastructure spending categorized by year.',
  table: {
    columns: ['Year', 'Number of Contracts', 'Total Value'],
    rows: yearRows.map(([y, d]) => [String(y), String(d.count), formatInr(d.value)]),
  },
});

// ── Write output ───────────────────────────────────────────────────────
fs.writeFileSync(OUTPUT_FILE, JSON.stringify(faqs, null, 2), 'utf8');
console.log(`\n✅ Generated ${faqs.length} FAQ entries → ${OUTPUT_FILE}`);
